lib/providers/api_client.dart

```diff
@@ -0,0 +1,149 @@
+        headers: response.headers,
+        reasonPhrase: response.reasonPhrase,
+        persistentConnection: response.persistentConnection,
+        contentLength: response.contentLength,
+        isRedirect: response.isRedirect,
+        request: response.request,
+      );
+    });
+  }
+
+  // Probably overkill.
+  @override
+  void close() {
+    _client.close();
+  }
+
+  // BaseClient.get(), but with response processing and error handling.
+  Future<Map<String, dynamic>> getJson(Uri url, {Map<String, String>? headers}) async {
+    _logRequest('Request', 'GET $url', null, headers, null);
+    return _parseResponse(
+        () async => await super.get(url, headers: await _authIfNeeded(url, headers)).timeout(RescadoConstants.timeout),
+      );
+  }
+
+  // BaseClient.post(), but with response processing and error handling.
+  Future<Map<String, dynamic>> postJson(Uri url, {Map<String, String>? headers, Object? body, Encoding? encoding}) async {
+    _logRequest('Request', 'POST $url', null, headers, body as String);
+    return _parseResponse(
+        () async => await super.post(url, headers: await _authIfNeeded(url, headers), body: body, encoding: encoding).timeout(RescadoConstants.timeout),
+      );
+  }
+
+  // BaseClient.put(), but with response processing and error handling.
+  Future<Map<String, dynamic>> putJson(Uri url, {Map<String, String>? headers, Object? body, Encoding? encoding}) async {
+    _logRequest('Request', 'PUT $url', null, headers, body as String);
+    return _parseResponse(
+        () async => await super.put(url, headers: await _authIfNeeded(url, headers), body: body, encoding: encoding).timeout(RescadoConstants.timeout),
+      );
+  }
+
+  // BaseClient.patch(), but with response processing and error handling.
+  Future<Map<String, dynamic>> patchJson(Uri url, {Map<String, String>? headers, Object? body, Encoding? encoding}) async {
+    _logRequest('Request', 'PATCH $url', null, headers, body as String);
+    return _parseResponse(
+        () async => await super.patch(url, headers: await _authIfNeeded(url, headers), body: body, encoding: encoding).timeout(RescadoConstants.timeout),
+      );
+  }
+
+  // BaseClient.delete(), but with response processing and error handling.
+  Future<Map<String, dynamic>> deleteJson(Uri url, {Map<String, String>? headers, Object? body, Encoding? encoding}) async {
+    _logRequest('Request', 'DELETE $url', null, headers, body as String);
+    return _parseResponse(
+        () async => await super.delete(url, headers: await _authIfNeeded(url, headers), body: body, encoding: encoding).timeout(RescadoConstants.timeout),
+      );
+  }
+
+  // If the url is not a public url, add the authorization header.
+  Future<Map<String, String>?> _authIfNeeded(Uri url, Map<String, String>? headers) async {
+    if (!url.path.contains('/auth/')) {
+      headers ??= {};
+      headers[HttpHeaders.authorizationHeader] = 'Bearer ${await RescadoStorage.getToken()}';
```

Comment:
Is `SharedPreferences.getInstance()` not a heavy operation every time /auth is called?

Reply:
```dart
  /// Loads and parses the [SharedPreferences] for this app from disk.
  ///
  /// Because this is reading from disk, it shouldn't be awaited in
  /// performance-sensitive blocks.
  static Future<SharedPreferences> getInstance() async {
    if (_completer == null) {
      final Completer<SharedPreferences> completer =
          Completer<SharedPreferences>();
      try {
        final Map<String, Object> preferencesMap =
            await _getSharedPreferencesMap();
        completer.complete(SharedPreferences._(preferencesMap));
      } on Exception catch (e) {
        // If there's an error, explicitly return the future with an error.
        // then set the completer to null so we can retry.
        completer.completeError(e);
        final Future<SharedPreferences> sharedPrefsFuture = completer.future;
        _completer = null;
        return sharedPrefsFuture;
      }
      _completer = completer;
    }
    return _completer!.future;
  }
  ```

tl;dr a whole lot of checks the first time but once done, it keeps a reference to itself and just returns that, which is instant. aka singleton. 🙂

Good remark though. I am creating a cached version in a provider, so the step to retrieve the data from disk every time (which will be more time consuming than from memory) can potentially be skipped. Silly to read the token from disk every time. Better to keep it in memory too.

- [ ] Make version of RescadoStorage with caching

Reply:
Done. Was replaced by `DeviceStorage`.